internal/attestations/authorization.go

```diff
@@ -0,0 +1,174 @@
+	ErrAuthorizationNotFound = errors.New("requested authorization not found")
+)
+
+// Authorization is a lightweight record of a detached authorization in a gittuf
+// repository. It is meant to be used as a "predicate" in an in-toto
+// attestation.
+type Authorization struct {
+	TargetRef    string `json:"targetRef"`
+	FromTargetID string `json:"fromTargetID"`
+	ToTargetID   string `json:"toTargetID"`
+}
+
+// NewAuthorizationAttestation creates a new authorization for the provided
+// information. The authorization is embedded in an in-toto "statement" and
+// returned with the appropriate "predicate type" set.
+func NewAuthorizationAttestation(targetRef, fromTargetID, toTargetID string) (*ita.Statement, error) {
+	predicate := &Authorization{
+		TargetRef:    targetRef,
+		FromTargetID: fromTargetID,
+		ToTargetID:   toTargetID,
+	}
+
+	predicateBytes, err := json.Marshal(predicate)
+	if err != nil {
+		return nil, err
+	}
+
+	predicateInterface := &map[string]any{}
+	if err := json.Unmarshal(predicateBytes, predicateInterface); err != nil {
+		return nil, err
+	}
+
+	predicateStruct, err := structpb.NewStruct(*predicateInterface)
+	if err != nil {
+		return nil, err
+	}
+
+	return &ita.Statement{
+		Type: ita.StatementTypeUri,
+		Subject: []*ita.ResourceDescriptor{
+			{
+				Digest: map[string]string{digestGitCommitKey: toTargetID},
+			},
+		},
+		PredicateType: AuthorizationPredicateType,
+		Predicate:     predicateStruct,
+	}, nil
+}
+
+// SetAuthorizationAttestation writes the new authorization attestation to the
+// object store and tracks it in the current attestations state.
+func (a *Attestations) SetAuthorizationAttestation(repo *git.Repository, env *sslibdsse.Envelope, refName, fromID, toID string) error {
+	if err := validateAuthorization(env, refName, fromID, toID); err != nil {
+		return err
+	}
+
+	envBytes, err := json.Marshal(env)
+	if err != nil {
+		return err
+	}
+
+	blobID, err := gitinterface.WriteBlob(repo, envBytes)
+	if err != nil {
+		return err
+	}
+
+	if a.authorizations == nil {
+		a.authorizations = map[string]plumbing.Hash{}
+	}
+
+	a.authorizations[AuthorizationPath(refName, fromID, toID)] = blobID
+	return nil
+}
+
+// RemoveAuthorizationAttestation removes a set authorization attestation
+// entirely. The object, however, isn't removed from the object store as prior
+// states may still need it.
+func (a *Attestations) RemoveAuthorizationAttestation(refName, fromID, toID string) error {
+	authPath := AuthorizationPath(refName, fromID, toID)
+	if _, has := a.authorizations[authPath]; !has {
+		return ErrAuthorizationNotFound
+	}
+
+	delete(a.authorizations, authPath)
+	return nil
+}
+
+// GetAuthorizationAttestationFor returns the requested authorization
+// attestation (with its signatures).
+func (a *Attestations) GetAuthorizationAttestationFor(repo *git.Repository, refName, fromID, toID string) (*sslibdsse.Envelope, error) {
```

Comment:
We should add documentation about merge branches and the expected behavior of attestations.

```mermaid
    gitGraph
       commit id: "A"
       branch develop
       commit id:"B"
       commit id:"C"
       checkout main
       merge develop id:"D"
       commit id:"E"
```

e.g. If there's an attestation added to `develop:B->C`, this means that when it's merged to `main` the attestation won't be searchable via `main`, even though the commits are now part of the `main` branch. It makes sense when you realize that Authorization attestations were created w.r.t. a particular ref, but that may not be something that's immediately obvious.

Reply:
I _think_ that's fine? The authorization is for `develop` to move from B to C. On the `main` branch, the authorization (if one's needed) is to move from `A` to `D`. For a parallel, if we map "authorization" to PR approval, we care that the PR moving `main` from A to D was approved and less so a sub-PR to the feature branch?

Reply:
Note that this is _strictly_ for authorizations as they work now. Perhaps it needs to be called `reference-state-authorization` or something to make it more clear?

Reply:
https://github.com/gittuf/gittuf/pull/195

I've documented this better here, it might help clarify why we may not want this to move with changes being merged / subsumed into other branches.